frontend/src/metabase-lib/v2/types.ts

```diff
@@ -0,0 +1,3 @@
+export type Query = unknown;
```

Comment:
In v2, the query object should be totally opaque for the FE, i.e. it'd be _really_ hard to work with directly without using lib functions. I'm not sure how to properly type to indicate it's not meant to be accessed directly but to also make sure we're passing around expected things

FWIW, here's how a v2 query looks like on the FE right now:

<img width="559" alt="CleanShot 2023-03-22 at 18 33 59@2x" src="https://user-images.githubusercontent.com/17258145/227003810-9e52f469-701d-4bd2-9ab9-975b7933d860.png">

Reply:
> make sure we're passing around expected things

Are these expected things context dependent? I.e., would it be intractable to encode the requirements into opaque types?

Reply:
@metamben I don't think typing e.g. a query object returned from MLv2 would make sense. I just want TypeScript to make sure that when I call `limit(query)`, it'd expect `query` to be what `question._getMLv2Query` returns

Reply:
Yeah, there's a trick to do that in TS. It makes distinct, "unforgeable" types that can only be created in the same file. 

We don't need to do it in this PR, but we can easily make it so all the TS wrappers in `frontend/src/metabase-lib/` that call CLJC things return these opaque types. then TS will check that eg. what you pass to `Lib.limit(q, 10)` must be the opaque `Query` (or whatever) type returned by other MLv2 functions. (Of course you can always `as unknown as Query` in TS but you'll get the bugs you deserve in that case.)

Reply:
Reference: https://calebmer.com/tangents/opaque-types-in-typescript.html
Example in a pending PR: https://github.com/metabase/metabase/blob/e0a93a89068be2ba688dbeadb44505d15056ab7f/frontend/src/metabase-lib/Question.ts#L114

Reply:
Agree with @bshepherdson , we could use opaque TS types for this. `unknown` has downsides as anything satisfies `unknown`

Reply:
[`Query` is now opaque](https://github.com/metabase/metabase/blob/c99ca5ac4fcd847c8bc8f77225dbd9dbba367211/frontend/src/metabase-lib/types.ts#L6). Passing another opaque type instead of a `Query` would still fail. Only two ways to bypass it are `@ts-ignore` or `as Query` which should be normally caught during the PR review